buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginPropertiesExtension.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.gradle.plugin;
+
+import org.gradle.api.Project;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * A container for plugin properties that will be written to the plugin descriptor, for easy
+ * manipulation in the gradle DSL.
+ */
+public class PluginPropertiesExtension {
+    private String name;
+
+    private String version;
+
+    private String description;
+
+    private String classname;
+
+    /** Other plugins this plugin extends through SPI */
+    private List<String> extendedPlugins = new ArrayList<>();
+
+    private boolean hasNativeController;
+
+    /** Indicates whether the plugin jar should be made available for the transport client. */
+    private boolean hasClientJar;
+
+    /** True if the plugin requires the elasticsearch keystore to exist, false otherwise. */
+    private boolean requiresKeystore;
+
+    /** A license file that should be included in the built plugin zip. */
+    private File licenseFile;
+
+    /**
+     * A notice file that should be included in the built plugin zip. This will be
+     * extended with notices from the {@code licenses/} directory.
+     */
+    private File noticeFile;
+
+    private final Project project;
+
+    public PluginPropertiesExtension(Project project) {
+        this.name = project.getName();
+        this.version = project.getVersion().toString();
```

Comment:
I just realized this now, this is a problem present in the old code as  well. 
The way we assign this default makes it prone to ordering. 
If the extension is created before `project.version` is assigned we won't pick it up. 
We should have a null check in the getter instead  and return`project.version` if `version` is `null`
This is fine also in a subsequent PR.

Reply:
I can update the `PluginPropertiesExtension.getVersion()` to return `project.getVersion().toString()` if `version` is null :+1:

Reply:
It is not possible for version to be null. VersionProperties statically loads the version. BuildPlugin   is initialized before any PluginPropertiesExtension, since it is applied by PluginBuildPlugin, which extends BuildPlugin. Adding null checks all over for version will only make the code more complicated to understand with no actual additional safety.

Reply:
We do this early on in the root level `build.gradle` 
```
allprojects {
  group = 'org.elasticsearch'
  version = VersionProperties.elasticsearch
  description = "Elasticsearch subproject ${project.path}"
}
```
So indeed it's not possible for `project.version` to be null in _our_ build, 
but all bets are off for external plugins, there's no guarantee that it won't be null. 
We might not want to be lenient here and throw an exception if the version is null, but I still think we should do it rather than leave it be an NPE. 
Alternatively, we could hard code this to `VersionProperties.elasticsearch` as we only test building plugins where the version of build-tools matches that of elasticsearch. 
External plugins can then have whatever project.version they like. 
What do you think of this approach @rjernst